### PR TITLE
Add initial LoongArch64 CPU support

### PR DESCRIPTION
**Pull request by liushuyu**
This pull request ports the D runtime to LoongArch64 CPU architecture.

LoongArch64 is a new CPU architecture and has been supported by GCC (since GCC 13) and LLVM/Clang (since LLVM 16).

You can find more information about this architecture here: https://loongson.github.io/LoongArch-Documentation/README-EN.html

### CODE REVIEW

File: compiler/src/dmd/cond.d [hunk 1]
```diff
@@ -691,6 +691,7 @@ extern (C++) final class VersionCondition : DVCondition
             case "LDC":
             case "linux":
             case "LittleEndian":
+            case "LoongArch64":
```
Comment (thewilsonator) [inline compiler/src/dmd/cond.d:694]: this is missing a test for it not being a user (command line) settable version

Comment (liushuyu) [reply compiler/src/dmd/cond.d:694]: Done. Tests added.

**Comment by liushuyu** [comment]
Cross-reference:

- LDC: https://github.com/ldc-developers/ldc/pull/4500
- GDC: https://gcc.gnu.org/pipermail/gcc-patches/2023-September/631241.html

File: compiler/src/dmd/cond.d [hunk 2]
```diff
@@ -691,6 +691,9 @@ extern (C++) final class VersionCondition : DVCondition
...
```
Comment (JohanEngelen) [inline compiler/src/dmd/cond.d:694]: perhaps also add `LoongArch32` in this PR, to reserve the version identifier for future use in druntime.

Comment (liushuyu) [reply compiler/src/dmd/cond.d:695]: Done.

File: druntime/src/core/threadasm.S [hunk 3]
```diff
@@ -229,6 +229,87 @@ fiber_switchContext:
 
     jr $ra // return
 
+#elif defined(__loongarch64)
+/************************************************************************************
+ * LoongArch64 ASM BITS
+ ************************************************************************************/
+
```
Comment (JohanEngelen) [inline druntime/src/core/threadasm.S:236]: maybe add a link to ELF ABI documentation: https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html

Comment (liushuyu) [reply druntime/src/core/threadasm.S:236]: Done. The link to the documentation is added below the "Parameters" section of the comment.

**Comment by WalterBright** [comment]
I'm puzzled. There's LoongArch64, LoongArch32, LoongArch_HardFloat, and LoongArch_SoftFloat. But the only one that is used is LoongArch64. I suggest not defining the others until there's a use case (otherwise we could define all kinds of architectures that we don't support, which is pointless).

**Comment by WalterBright** [comment]
Other than that, I approve this.

**Comment by xen0n** [comment]
Yeah currently Linux only supports the 64-bit userland, and 32-bit LoongArch support is patchy right now. I'd agree with only defining LoongArch64 (and maybe the hard-float symbol) until use cases for the rest arise.

**Comment by ibuclaw** [comment]
Something should perhaps be predefined though when the base architecture is `!TARGET_64BIT` in gdc/ldc backend, otherwise there is no indication that compiler is generating code for a LoongArch processor.

**Comment by liushuyu** [comment]
> Something should perhaps be predefined though when the base architecture is `!TARGET_64BIT` in gdc/ldc backend, otherwise there is no indication that compiler is generating code for a LoongArch processor.

I agree with that. Currently, GDC and LDC implementations for LoongArch32/64 architectures do use all those version identifiers.

**Comment by liushuyu** [comment]
I have rebased this pull request against the latest `master` branch.

**Comment by liushuyu** [comment]
Hi,

What is the status of this pull request? Should I do something or tag someone for review?

**Review by JohanEngelen — APPROVED** [review]

**Pull request merged**